rpm/0071-sailfishos-webrtc-Implement-video-capture-module.-JB.patch

```diff
@@ -0,0 +1,797 @@
++{
++VideoCaptureModule::DeviceInfo*
++VideoCaptureImpl::CreateDeviceInfo()
++{
++    // The caller doesn't check for return value, so I can't Init() here
++    return new videocapturemodule::DeviceInfoSFOS();
++}
++
++DeviceInfoSFOS::DeviceInfoSFOS()
++    : DeviceInfoImpl()
++{
++    Init();
++}
++
++int32_t DeviceInfoSFOS::Init()
++{
++    cameraManager = amber_camera_manager();
++    // Initialize parent class member
++    _lastUsedDeviceName = (char *)malloc(1);
++    _lastUsedDeviceName[0] = 0;
++    return 0;
++}
++
++DeviceInfoSFOS::~DeviceInfoSFOS()
++{
++}
++
++uint32_t DeviceInfoSFOS::NumberOfDevices()
++{
++    LOG_F(LS_VERBOSE);
++
++    if (cameraManager) {
++        cameraList.clear();
++        for (int i = 0; i < cameraManager->getNumberOfCameras(); i++) {
++            amber::media::CameraInfo info;
++            if (cameraManager->getCameraInfo(i, info)) {
++                cameraList.push_back(info);
++            }
++        }
++        // Put front cameras at the top of the list, they are most often used during video chat.
++        std::sort(cameraList.begin(), cameraList.end(),
++                [](const amber::media::CameraInfo& c1, const amber::media::CameraInfo& c2) {
++                    bool c1front = c1.facing == amber::media::AMBER_CAMERA_FACING_FRONT;
++                    bool c2front = c2.facing == amber::media::AMBER_CAMERA_FACING_FRONT;
++                    return c1front > c2front;
++            });
++        return cameraList.size();
++    }
++    return 0;
++}
++
++int32_t DeviceInfoSFOS::GetDeviceName(
++        uint32_t deviceNumber,
++        char* deviceNameUTF8,
++        uint32_t deviceNameLength,
++        char* deviceUniqueIdUTF8,
++        uint32_t deviceUniqueIdUTF8Length,
++        char* productUniqueIdUTF8,
++        uint32_t productUniqueIdUTF8Length,
++        pid_t* /*pid*/)
++{
++    if (deviceNumber < cameraList.size()) {
++        amber::media::CameraInfo info = cameraList.at(deviceNumber);
++        strncpy(deviceNameUTF8, info.name.c_str(), deviceNameLength);
++        strncpy(deviceUniqueIdUTF8, info.id.c_str(), deviceUniqueIdUTF8Length);
++        strncpy(productUniqueIdUTF8, info.provider.c_str(), productUniqueIdUTF8Length);
++        return 0;
++    }
++
++    return -1;
++}
++
++int32_t DeviceInfoSFOS::CreateCapabilityMap(const char* deviceUniqueIdUTF8)
++{
++    const int32_t deviceUniqueIdUTF8Length =
++                            (int32_t) strlen((char*) deviceUniqueIdUTF8);
++    if (deviceUniqueIdUTF8Length > kVideoCaptureUniqueNameLength) {
++        LOG_F(LS_ERROR) << "Device name too long";
++        return -1;
++    }
++
++    FillCapabilities(deviceUniqueIdUTF8);
++
++    // Store the new used device name. The parent class needs this for some reason
++    _lastUsedDeviceNameLength = deviceUniqueIdUTF8Length;
++    _lastUsedDeviceName = (char*) realloc(_lastUsedDeviceName,
++                                                   _lastUsedDeviceNameLength + 1);
```

Comment:
Could this be a std::string?

Reply:
This is the member of the parent class. Other capture implementations also fill it. Do you think I should replace it everywhere to std::string? I'd happy to, but it will make porting to future versions of gecko a bit harder.

Reply:
No, don't do that. I should have found the definition.